src/NeoSharp.Core/Messaging/MessageHandlerProxy.cs

```diff
@@ -1,153 +1,64 @@
 using System;
 using System.Collections.Generic;
-using System.Linq;
-using System.Linq.Expressions;
-using System.Reflection;
 using System.Threading.Tasks;
-using NeoSharp.Core.DI;
 using NeoSharp.Core.Logging;
 using NeoSharp.Core.Network;
 
 namespace NeoSharp.Core.Messaging
 {
     public class MessageHandlerProxy : IMessageHandler<Message>
     {
-        #region Variables
-
-        private readonly IContainer _container;
+        #region Private fields 
+        private readonly IEnumerable<IMessageHandler> _messageHandlers;
         private readonly ILogger<MessageHandlerProxy> _logger;
-        private readonly IReadOnlyDictionary<Type, Delegate> _messageHandlerInvokers;
-
         #endregion
 
-        /// <summary>
-        /// Constructor
-        /// </summary>
-        /// <param name="container">Container</param>
-        /// <param name="messageHandlerTypes">MessageHandler types</param>
-        /// <param name="logger">Logger</param>
-        public MessageHandlerProxy(IContainer container, IEnumerable<Type> messageHandlerTypes, ILogger<MessageHandlerProxy> logger)
+        #region Constructor 
+        public MessageHandlerProxy(
+            IEnumerable<IMessageHandler> messageHandlers, 
+            ILogger<MessageHandlerProxy> logger)
         {
-            _container = container;
+            _messageHandlers = messageHandlers;
             _logger = logger;
-            _messageHandlerInvokers = messageHandlerTypes
-                .Select(CreateMessageHandlerInvoker)
-                .ToDictionary(x => x.MessageType, x => x.MessageHandlerInvoker);
         }
+        #endregion
 
-        /// <summary>
-        /// Handle message
-        /// </summary>
-        /// <param name="message">Message</param>
-        /// <param name="sender">Sender</param>
-        /// <returns>Task</returns>
-        public Task Handle(Message message, IPeer sender)
+        #region IMessageHandler implementation 
+        public async Task Handle(Message message, IPeer sender)
         {
             if (message == null) throw new ArgumentNullException(nameof(message));
 
-            var messageType = message.GetType();
-            var messageHandler = ResolveMessageHandler(messageType);
-            var messageHandlerName = messageHandler.GetType().Name;
-            var startedAt = LogMessageHandlingStart(messageHandlerName);
-            var messageHandlerInvoker = GetMessageHandlerInvoker(messageType);
-            var handleMessageTask = (Task)messageHandlerInvoker.DynamicInvoke(messageHandler, message, sender);
-
-            return handleMessageTask.ContinueWith(t =>
+            var isHandled = false;
+            foreach (var handler in this._messageHandlers)
```

Comment:
We cannot iterate through all message handler for every incoming message. It will be, for sure, a bottleneck.

Reply:
of course not .. will be a linq query similar of what we had that, from all the MessageHadlers will return the only one that satisfy the condition.

Reply:
it is still iterating even being hidden in LINQ

Reply:
we were already doing the same ... we where receiving a list and doing that ... this is just a clean way without using Reflection to create the objects.

Reply:
What we are doing is we ask DI container to create the __type safe version__ (that has Handle method with the first argument of _actual_ Message type) of a message handler __per message__.

What this change introduces is it ask DI container to create the __type unsafe version__ (that has Handle method with the first argument of _generic_ Message type) of a message handler as __singleton__ (single message handler will handle all compatible messages while the app is running) and, in addition, it searches per message through message handler instances to discover who first will be able to handle the message.

Reply:
I think Simple Injector allow that we can have Type Safe list injected in the constructor. I could not find a constructive example, just the documentation on their website

https://simpleinjector.readthedocs.io/en/latest/advanced.html?highlight=Collection

I will create a Issue requesting that this could be refactor and how.

Reply:
issue #427 added

Reply:
Made a small refactor on the Handle method where the Foreach is removed.

I'm not a fan of the if statement
if (messageHandler == null)
            {
                this._logger.LogError($"could not find the handler for the message type {message.GetType()}");
                return;
            }
but I think that for now, until this method is not stable we can live with this. The LINQ query should return Single() because all the possible Messages should have a MessageHandler, if not, the developer didn't made is work well.